server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/GitMirror.java

```diff
@@ -333,4 +375,320 @@ private Git openGit(File workDir) throws Exception {
+     */
+    private static String sanitizeText(String text) {
+        if (text.indexOf('\r') >= 0) {
+            text = CR.matcher(text).replaceAll("");
+        }
+        if (!text.isEmpty() && !text.endsWith("\n")) {
+            text += "\n";
+        }
+        return text;
+    }
+
+    private ObjectId commit(org.eclipse.jgit.lib.Repository gitRepository, DirCache dirCache,
+                            ObjectId headCommitId, Revision localHead) throws IOException {
+        try (ObjectInserter inserter = gitRepository.newObjectInserter()) {
+            // flush the current index to repository and get the result tree object id.
+            final ObjectId nextTreeId = dirCache.writeTree(inserter);
+            // build a commit object
+            final PersonIdent personIdent =
+                    new PersonIdent(MIRROR_AUTHOR.name(), MIRROR_AUTHOR.email(),
+                                    System.currentTimeMillis() / 1000L * 1000L, // Drop the milliseconds
+                                    0);
+
+            final CommitBuilder commitBuilder = new CommitBuilder();
+            commitBuilder.setAuthor(personIdent);
+            commitBuilder.setCommitter(personIdent);
+            commitBuilder.setTreeId(nextTreeId);
+            commitBuilder.setEncoding(UTF_8);
+            commitBuilder.setParentId(headCommitId);
+
+            final String summary = "Mirror '" + localRepo().name() + "' at " + localHead +
+                                   " to the repository '" + remoteRepoUri() + '#' + remoteBranch() + "'\n";
+            logger.info(summary);
+            commitBuilder.setMessage(summary);
+
+            final ObjectId nextCommitId = inserter.insert(commitBuilder);
+            inserter.flush();
+            return nextCommitId;
+        }
+    }
+
+    static void updateRef(org.eclipse.jgit.lib.Repository jGitRepository, RevWalk revWalk,
+                          String ref, ObjectId commitId) throws IOException {
+        final RefUpdate refUpdate = jGitRepository.updateRef(ref);
+        refUpdate.setNewObjectId(commitId);
+
+        final Result res = refUpdate.update(revWalk);
+        switch (res) {
+            case NEW:
+            case FAST_FORWARD:
+                // Expected
+                break;
+            default:
+                throw new StorageException("unexpected refUpdate state: " + res);
+        }
+    }
+
+    private static final class InsertText extends PathEdit {
+        private final ObjectInserter inserter;
+        private final String text;
+
+        InsertText(String entryPath, ObjectInserter inserter, String text) {
+            super(entryPath);
+            this.inserter = inserter;
+            this.text = text;
+        }
+
+        @Override
+        public void apply(DirCacheEntry ent) {
+            try {
+                ent.setObjectId(inserter.insert(Constants.OBJ_BLOB, text.getBytes(UTF_8)));
```

Comment:
I thought we need to convert to `String` first because we need to use the pretty printer. If we use bytes, the mirror repository would see something like
`{"a": {"foo":"bar"}}` instead of
```
{
  "a" :  {
    "foo": "bar"
  }
}
```

Reply:
I see, I think this is reasonable:
- For remote git repositories, we pretty print
- For local storage, we save the compact json